src/v/cluster/leader_routing_frontend.h

```diff
@@ -0,0 +1,248 @@
+// Copyright 2023 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#pragma once
+
+#include "cluster/logger.h"
+#include "cluster/members_table.h"
+#include "cluster/metadata_cache.h"
+#include "cluster/partition_leaders_table.h"
+#include "cluster/partition_manager.h"
+#include "cluster/shard_table.h"
+#include "cluster/types.h"
+#include "model/metadata.h"
+#include "model/timeout_clock.h"
+#include "rpc/connection_cache.h"
+#include "seastarx.h"
+#include "vformat.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/sharded.hh>
+
+namespace rpc {
+class connection_cache;
+} // namespace rpc
+
+namespace cluster {
+
+// Encapsulates a frontend that aims to direct RPCs to a specific NTP's leader.
+//
+// When a client invokes process_or_dispatch() on a remote node, the frontend
+// redirects the request to the service located on the broker that hosts the
+// NTP's leader.
+//
+// When a client is colocated with the leader, the frontend calls
+// find_shard_and_process() directly, calling into subclasses' process()
+// implementations.
+template<typename req_t, typename resp_t, typename proto_t>
+class leader_routing_frontend {
+public:
+    using duration = model::timeout_clock::duration;
+    leader_routing_frontend(
+      ss::sharded<cluster::partition_manager>&,
+      ss::sharded<cluster::shard_table>&,
+      ss::sharded<cluster::metadata_cache>&,
+      ss::sharded<rpc::connection_cache>&,
+      ss::sharded<partition_leaders_table>&,
+      model::topic_namespace,
+      model::node_id self,
+      int16_t,
+      std::chrono::milliseconds);
+
+    // Sends the given request to the leader, or processes it on the
+    // appropriate shard if this node is the leader.
+    virtual ss::future<resp_t>
+    process_or_dispatch(req_t req, model::partition_id, duration timeout);
+
+    // Finds the shard appropriate for the NTP and processes the given request,
+    // expecting the leader is on this node.
+    virtual ss::future<resp_t>
+    find_shard_and_process(req_t req, model::partition_id, duration timeout);
+
+protected:
+    // Sends the given request on the given client protocol with the given
+    // timeout.
+    virtual ss::future<result<rpc::client_context<resp_t>>>
```

Comment:
wondering if you considered a direct templated function instead of a virtual dispatch.. something like..

template<class LocalFunc, class RemoteFunc, class Req, class Resp>
ss::future\<Resp\> local_or_remote_dispatch(ntp, req, LocalFunc&&, RemoteFunc&&);

Reply:
Yeah switched over to templates and it ends up similar to what you're proposing. I wrapped the local and remote functions in their own class though.